src/main/java/net/minecraftforge/common/extensions/IForgeJukeboxBlock.java

```diff
@@ -0,0 +1,94 @@
+import net.minecraft.world.level.block.JukeboxBlock;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.JukeboxBlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Extension-Interface providing methods for various jukebox activiites that are currently hardcoded to only the vanilla jukebox.
+ * Any {@link Block} implementing will be treated as if it was a jukebox.
+ *
+ * @see JukeboxBlock
+ */
+public interface IForgeJukeboxBlock
+{
+
+    /**
+     * Called by {@link RecordItem} to insert a record into the jukebox.
+     * The actual action taken by the jukebox is up to it and not the record.
+     *
+     * @param state The blockstate of the jukebox.
+     * @param level The level of the jukebox.
+     * @param pos The position of the jukebox.
+     * @param entity The entity inserting into jukebox. This can be null.
+     * @param stack The itemstack being inserted.
+     */
+    default void insertRecord(BlockState state, Level level, BlockPos pos, @Nullable Entity entity, ItemStack stack)
+    {
+        if (this instanceof JukeboxBlock jukebox)
+            jukebox.setRecord(entity, level, pos, state, stack);
+    }
+
+    /**
+     * Gets the {@link ItemStack} of the current record in the jukebox.
+     * The current record can be empty or any item. In vanilla, this can only be records.
+     *
+     * @param state The blockstate of the jukebox.
+     * @param level The level of the jukebox.
+     * @param pos The position of the jukebox.
+     * @return the current record or empty.
+     */
+    default ItemStack getRecord(BlockState state, Level level, BlockPos pos)
+    {
+        if (level.getBlockEntity(pos) instanceof JukeboxBlockEntity jukebox)
+            return jukebox.getRecord();
+        else
+            return ItemStack.EMPTY;
+    }
+
+    /**
+     * Checks if an entity is allowed to dance when in proximity to the jukebox.
+     * The actual behavior is controlled by the entity.
+     * Right now only the vanilla parrot/allay currently use this.
+     *
+     * @param state The blockstate of the jukebox.
+     * @param entity The entity being checked.
+     * @return true if the entity can dance.
+     */
+    default boolean canDance(BlockState state, Entity entity)
+    {
+        return entity instanceof Allay || entity instanceof Parrot;
+    }
+
+    /**
+     * Gets if the jukebox currently has a record inside it.
+     *
+     * @param state The blockstate of the jukebox.
+     * @param level The level of the jukebox.
+     * @param pos The position of the jukebox.
+     * @return true if jukebox has a record.
+     */
+    default boolean hasRecord(BlockState state, Level level, BlockPos pos)
+    {
+        if (this instanceof JukeboxBlock)
+            return !state.getValue(JukeboxBlock.HAS_RECORD);
+        else
+            return false;
```

Comment:
As a convenience, perhaps this could delegate to `getRecord()` if not a vanilla jukebox?

Reply:
What it could do is call `this.getRecord()`, then return whether that returned an non-empty item. That way custom jukebox blocks that don't have a cached has-record value could only implement `getRecord()`, and have `hasRecord()` do the right thing.

Reply:
The value of `#getRecord` would be called twice, then. If that strategy was adopted, then it would make no sense to have a `#hasRecord` at all and just implement the logic into a single method.

This is not to mention I can see cases where somebody decides to check the item and wonder why the client and server are desynced to some capacity. To implement that more complex logic, they would need to properly handle any tags which may change the logic. They would be overriding both methods anyways in that case. As such, I don't think it makes sense to do so.